Juan Miguel Villegas Yeguas---Teoría de Números y Criptografía---Ejercicio RSA

Mi DNI es 77021622, lo partimos en dos bloques

In [1]:
bloque1=7702
bloque2=1622

Ambos bloques son mayores que 1000, luego continuamos, como los dos bloques son pares
nuestro p y q serán los primos más cercanos. 

In [2]:
p=next_prime(bloque1)
q=next_prime(bloque2)
print (p,q)

7703 1627


En mi caso p=7703 y q=1627

Consideramos n=p*q y buscamos el menor primo mayor que 11 que sea primo relaticos con 
phi(n)

In [3]:
n=p*q
phi=n+1-(p+q)
print (n,phi)

12532781 12523452


In [4]:
11.gcd(phi)

1

Tenemos que n=12532781 y phi(n)=12523452, hemos comprobado que 11 es primo relativo con phi, tomamos e=11

In [5]:
e=11

Ahora calculamos el inverso de e modulo phi(n)

In [6]:
d=inverse_mod(e,phi)
print(d)

3415487


Comprobamos que lo hemos calculado bien

In [7]:
(e*d)%phi

1

Comenzamos con los apartados el ejercicio

#Apartado 1#

Introducimos el mensaje y lo ciframos con la función de cifrado de RSA

In [8]:
m=0xcafe

Definimos la función de cifrado de RSA

In [9]:
def RSA_cifrado(m,e,n):
    cifrado=power_mod(m,e,n)
    return cifrado;

Nuestro criptograma viene dado por

In [10]:
clave=RSA_cifrado(m,e,n)
'0x'+clave.hex()

'0x1eefcd'

#Apartado 2#

Vamos a descifrar el criptograma anterior, para ello, definimos la función de descifrado de RSA

In [11]:
def RSA_descifrado(c,e,n):
    d=inverse_mod(e,euler_phi(n))
    descifrado=power_mod(c,d,n)
    return descifrado;

El mensaje original viene dado por

In [12]:
mensaje=RSA_descifrado(clave,e,n)
'0x'+mensaje.hex()

'0xcafe'

#Apartado 3#

Vamos a intentar factorizar n mediante el método P-1 de Pollard llegando como máximo a b=8

In [13]:
b=1
m=2
a=m.powermod(b,n)
(a-1).gcd(n)

1

Como vale el mcd de a-1 y n es 1 vamos probando con otro b

In [20]:
b+=1
a=a.powermod(b,n)
print(b,a)
print((a-1).gcd(n))

8 1606542
1


Como hemos llegado a b=8 y no hemos encontrado ningún factor, volvemos a hacer el mismo proceso con m=3

In [21]:
3.gcd(n)

1

Tomamos m=3 y repetimos el algoritmo

In [22]:
b=1
m=3
a=m.powermod(b,n)
(a-1).gcd(n)

1

In [29]:
b+=1
a=a.powermod(b,n)
print(b,a)
print((a-1).gcd(n))

8 7901858
1


Volvemos a hacer otra eleción de m con el mismo critero que antes y despues ejecutamos el algoritmo
hasta que encontremos un factor de n

In [30]:
4.gcd(n)

1

In [31]:
b=1
m=4
a=m.powermod(b,n)
(a-1).gcd(n)

1

In [38]:
b+=1
a=a.powermod(b,n)
print(b,a)
print((a-1).gcd(n))

8 1344186
1


Para agilizar el proceso implementemos el siguiente bucle

In [39]:
parada=false
for i in IntegerRange(5,300):
    if i.gcd(n)==1:
        m=i
        b=1
        a=m.powermod(b,n)
        if((a-1).gcd(n) !=1):
            print (i,a,b)
            print ((a-1).gcd(n))
            parada=true
        if parada==true:
            break
        for b in [2,3,4,5,6,7,8]:
            a=a.powermod(b,n)
            if (a-1).gcd(n)!=1:
                print(i,a,b)
                print('Uno de los factores de n es:') 
                print((a-1).gcd(n))
                parada=true
                break
        if parada==true:
            break

264 9425212 3
Uno de los factores de n es:
1627


In [40]:
factor1=(a-1).gcd(n)
factor2=n//factor1
print(factor1, factor2)
print('Hemos encotrado la descomposición de n en factores primos.')
n==factor1*factor2

1627 7703
Hemos encotrado la descomposición de n en factores primos.


True

#Apartado 4#

Vamos a factorizar n a partir de phi(n)

Consideramos el polinomio (x-p)(x-q)=x^2-(p+q)x+n,  como phi(n)= n+1-(p+q) tenemos (x-p)(x-q)=x^2+(phi(n)-n-1)x+n, luego p y q son las raices del polinomio anterior

In [41]:
x = var('x')
solve(x^2-(n+1-phi)*x+n, x)

[x == 1627, x == 7703]

#Apartado 5#

Factoricemos n a partir de e y d, lo haremos haciendo el mismo método que el ejercicio 3.5 de la teoría

In [42]:
k=e*d-1;
print(k)

37570356


In [43]:
for i in [2,3,5,7,11,13,17,19]:
    print(power_mod(i,k,n), i, k ,n)

1 2 37570356 12532781
1 3 37570356 12532781
1 5 37570356 12532781
1 7 37570356 12532781
1 11 37570356 12532781
1 13 37570356 12532781
1 17 37570356 12532781
1 19 37570356 12532781


In [44]:
k=k//2;
print(k)
for i in [2,3,5,7,11,13,17,19]:
    print(power_mod(i,k,n), i, k ,n)

18785178
1 2 18785178 12532781
1 3 18785178 12532781
1 5 18785178 12532781
1 7 18785178 12532781
1 11 18785178 12532781
1 13 18785178 12532781
1 17 18785178 12532781
1 19 18785178 12532781


In [45]:
k=k//2;
print(k)
for i in [2,3,5,7,11,13,17,19]:
    print(power_mod(i,k,n), i, k ,n)

9392589
4467741 2 9392589 12532781
4467741 3 9392589 12532781
12532780 5 9392589 12532781
1 7 9392589 12532781
8065040 11 9392589 12532781
12532780 13 9392589 12532781
4467741 17 9392589 12532781
1 19 9392589 12532781


Vemos que 2^9392589 mod n = 4467741, como no es uno, un factor de n viene dado por mcd(n,2^9392589-1 mod n)

In [46]:
p=n.gcd(4467740)
q=n//p
print(p,q)

7703 1627


Comprobamos que son los factores de n

In [47]:
n==p*q

True